numpy/fft/_pocketfft_umath.cpp

```diff
@@ -0,0 +1,437 @@
+ */
+
+/*
+ *  Main implementation file.
+ *
+ *  Copyright (C) 2004-2018 Max-Planck-Society
+ *  \author Martin Reinecke
+ */
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#define PY_SSIZE_T_CLEAN
+#include <assert.h>
+#include <Python.h>
+
+#include "numpy/arrayobject.h"
+#include "numpy/ufuncobject.h"
+
+#include "npy_config.h"
+
+#define POCKETFFT_NO_MULTITHREADING
+#include "pocketfft/pocketfft_hdronly.h"
+
+/*
+ * Transfer to and from a contiguous buffer.
+ * copy_input: copy min(nin, n) elements from input to buffer and zero rest.
+ * copy_output: copy n elements from buffer to output.
+ */
+template <typename T>
+static inline void
+copy_input(char *in, npy_intp step_in, size_t nin,
+           T buff[], size_t n)
+{
+    size_t ncopy = nin <= n ? nin : n;
+    char *ip = in;
+    size_t i;
+    for (i = 0; i < ncopy; i++, ip += step_in) {
+      buff[i] = *(T *)ip;
+    }
+    for (; i < n; i++) {
+      buff[i] = 0;
+    }
+}
+
+template <typename T>
+static inline void
+copy_output(T buff[], char *out, npy_intp step_out, size_t n)
+{
+    char *op = out;
+    for (size_t i = 0; i < n; i++, op += step_out) {
+        *(T *)op = buff[i];
+    }
+}
+
+/*
+ * Gufunc loops calling the pocketfft code.
+ */
+template <typename T>
+static void
+fft_loop(char **args, npy_intp const *dimensions, ptrdiff_t const *steps,
+         void *func)
+{
+    char *ip = args[0], *fp = args[1], *op = args[2];
+    size_t n_outer = (size_t)dimensions[0];
+    ptrdiff_t si = steps[0], sf = steps[1], so = steps[2];
+    size_t nin = (size_t)dimensions[1], nout = (size_t)dimensions[2];
+    ptrdiff_t step_in = steps[3], step_out = steps[4];
+    bool direction = *((bool *)func); /* pocketfft::FORWARD or BACKWARD */
+
+    assert (nout > 0);
+
+#ifndef POCKETFFT_NO_VECTORS
+    /*
+     * For the common case of nin >= nout, fixed factor, and suitably sized
+     * outer loop, we call pocketfft directly to benefit from its vectorization.
+     * (For nin>nout, this just removes the extra input points, as required;
+     * the vlen constraint avoids compiling extra code for longdouble, which
+     * cannot be vectorized so does not benefit.)
+     */
+    constexpr auto vlen = pocketfft::detail::VLEN<T>::val;
+    if (vlen > 1 && n_outer >= vlen && nin >= nout && sf == 0) {
+        std::vector<size_t> shape = { n_outer, nout };
+        std::vector<ptrdiff_t> strides_in = { si, step_in };
+        std::vector<ptrdiff_t> strides_out = { so, step_out};
+        std::vector<size_t> axes = { 1 };
+        pocketfft::c2c(shape, strides_in, strides_out, axes, direction,
+                       (std::complex<T> *)ip, (std::complex<T> *)op, *(T *)fp);
+        return;
+    }
+#endif
+    /*
+     * Otherwise, use a non-vectorized loop in which we try to minimize copies.
+     * We do still need a buffer if the output is not contiguous.
+     */
+    auto plan = pocketfft::detail::get_plan<pocketfft::detail::pocketfft_c<T>>(nout);
```

Comment:
@mreineck - for guarding against out-of-memory, getting the plan here would be relevant, but not the actual execution, correct?

Plus of course the call to the iterator `pocketfft:c2c` above.

These two get repeated twice more, but it should be easy to abstract that a bit if needed.